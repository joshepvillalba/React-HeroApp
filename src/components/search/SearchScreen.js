import React,{ useMemo } from 'react'
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm }from '../../hooks/useForm'
import { HeroCard } from '../hero/HeroCard';
import { getHeroByName } from '../selectors/getHeroByName';
import queryString from 'query-string';


export const SearchScreen = () => {

  const navigate = useNavigate();
  const localtion = useLocation();
  
  const { q = ''} = queryString.parse(localtion.search)

  const [formValue, handleInputChange] = useForm({
    searchtext: q ,
  })

  const { searchtext }= formValue;

  const data = useMemo(() =>  getHeroByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchtext }`)
  }


  return (
    <>
      <h1>Busqueda</h1>
      <hr/>

      <div className='row'>
        <div className='col-sm-12 col-md-3'>
          <h4>Buscar</h4>
          <hr/>

          <form onSubmit = { handleSearch }>

            <input type='text'
            placeholder='Buscar heroe'
            className='form-control'
            name='searchtext'
            autoComplete='off'
            value = { searchtext }
            onChange = { handleInputChange }/>
            
            <button type='submit' className='btn btn-primary btn-lg btn-block mt-1' style={{width: '100%'}}>Buscar...</button>

          </form>

        </div>
        
        <div className='col-sm-12 col-md-9 mt-2'>
          <h4>Resultados</h4>
          <hr/>
          
          {
                data.length >= 1 ? data.map( hero => <HeroCard key = {hero.id} { ...hero }  />) : data == []? <h3 className=' alert alert-info'>Buscar un heroe</h3>: <h3 className='alert alert-warning'>No se encontro el heroe: {q}</h3>
          }
        </div>

      </div>


    </>
  )
}
