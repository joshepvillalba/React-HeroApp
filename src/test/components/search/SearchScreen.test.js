import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))


describe('Pruebas en <SearchScreen />', () => { 

    test('Debe mostrar el componente con valores por defecto', () => { 

        const wrapper = mount( 
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>
        
        )

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert').text().trim() ).toBe('Buscar un heroe');

     })

     test('Debe de mostrar a batman y el input con el queryString', () => { 

        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
            )

            expect( wrapper ).toMatchSnapshot();
            expect( wrapper.find('input').prop('value') ).toBe('batman');
      })

      test('Debe de mostrar error si no se encuentra el heroe', () => { 

        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=123']}>
                <SearchScreen />
            </MemoryRouter>
            )

            expect( wrapper ).toMatchSnapshot();
            expect( wrapper.find('.alert').text().trim() ).toBe('No se encontro el heroe: 123');
      })


      test('Debe de llamar el navigate a la nueva pantalla', () => { 

        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
            );

            wrapper.find('input').simulate('change', {target:{name:'searchtext', value:'batman'}});
            wrapper.find('form').prop('onSubmit')({preventDefault: () => {}});

            expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
      })

 })