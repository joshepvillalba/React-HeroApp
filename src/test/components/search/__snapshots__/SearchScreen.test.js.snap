// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar a batman y el input con el queryString 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchtext"
          onChange={[Function]}
          placeholder="Buscar heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary btn-lg btn-block mt-1"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeInDown"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  className="text-muted"
                >
                  No hay mas personajes con este avatar
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más información
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar error si no se encuentra el heroe 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchtext"
          onChange={[Function]}
          placeholder="Buscar heroe"
          type="text"
          value="123"
        />
        <button
          className="btn btn-primary btn-lg btn-block mt-1"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <h3
        className="alert alert-warning"
      >
        No se encontro el heroe: 
        123
      </h3>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Debe mostrar el componente con valores por defecto 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchtext"
          onChange={[Function]}
          placeholder="Buscar heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary btn-lg btn-block mt-1"
          style={
            Object {
              "width": "100%",
            }
          }
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <h3
        className=" alert alert-info"
      >
        Buscar un heroe
      </h3>
    </div>
  </div>,
]
`;
